File: public_admin/server/ws_ticket/service.py
from __future__ import annotations

import hashlib
import secrets
from datetime import datetime, timedelta
from typing import Any

from .models import WsTicketClaims, WsTicketIssue


class WsTicketError(Exception):
    def __init__(self, code: str, message: str):
        super().__init__(message)
        self.code = code
        self.message = message


class WsTicketService:
    def __init__(self, repository, *, ttl_seconds: int = 45, logger: Any = None, diagnostics_policy: Any = None):
        self.repository = repository
        self.ttl_seconds = max(10, min(300, int(ttl_seconds or 45)))
        self.logger = logger
        self.diagnostics_policy = diagnostics_policy

    async def ensure_schema(self) -> None:
        if self.repository and hasattr(self.repository, "ensure_schema"):
            await self.repository.ensure_schema()

    async def issue(
        self,
        *,
        audience: str,
        subject: str,
        role: str = "",
        resource_type: str = "",
        resource_id: str = "",
        site: str = "",
        readonly: bool = False,
        claims: dict[str, Any] | None = None,
        client_ip: str = "",
        user_agent: str = "",
    ) -> WsTicketIssue:
        normalized_audience = self._normalize_required(audience, "audience").lower()
        normalized_subject = self._normalize_required(subject, "subject").lower()
        normalized_role = str(role or "").strip().lower()
        token = secrets.token_urlsafe(32)
        token_hash = self.hash_token(token)
        now = datetime.now().replace(microsecond=0)
        expires_at = now + timedelta(seconds=self.ttl_seconds)
        claim_model = WsTicketClaims(
            audience=normalized_audience,
            subject=normalized_subject,
            role=normalized_role,
            resource_type=str(resource_type or "").strip(),
            resource_id=str(resource_id or "").strip(),
            site=str(site or "").strip(),
            readonly=bool(readonly),
            claims=dict(claims or {}),
            issued_at=now,
            expires_at=expires_at,
        )
        await self.repository.insert_ticket(
            token_hash=token_hash,
            audience=claim_model.audience,
            subject=claim_model.subject,
            role=claim_model.role,
            resource_type=claim_model.resource_type,
            resource_id=claim_model.resource_id,
            site=claim_model.site,
            readonly=claim_model.readonly,
            claims=claim_model.claims,
            issued_at=now,
            expires_at=expires_at,
            client_ip=client_ip,
            user_agent=user_agent,
        )
        await self._record_event_safe(
            event_type="issue",
            code="ok",
            claims=claim_model,
            client_ip=client_ip,
            user_agent=user_agent,
        )
        return WsTicketIssue(
            ticket=token,
            token_hash=token_hash,
            claims=claim_model,
            expires_in=self.ttl_seconds,
        )

    async def consume(
        self,
        *,
        ticket: str,
        audience: str,
        consume_ip: str = "",
        consume_user_agent: str = "",
    ) -> WsTicketClaims:
        normalized_ticket = str(ticket or "").strip()
        normalized_audience = self._normalize_required(audience, "audience").lower()
        if not normalized_ticket:
            await self.record_event(
                event_type="reject",
                code="missing_ticket",
                audience=normalized_audience,
                consume_ip=consume_ip,
                user_agent=consume_user_agent,
            )
            raise WsTicketError("missing_ticket", "missing websocket ticket")
        row = await self.repository.consume_ticket(
            token_hash=self.hash_token(normalized_ticket),
            audience=normalized_audience,
            consume_ip=consume_ip,
            consume_user_agent=consume_user_agent,
        )
        if not row:
            await self.record_event(
                event_type="reject",
                code="invalid_ticket",
                audience=normalized_audience,
                consume_ip=consume_ip,
                user_agent=consume_user_agent,
            )
            raise WsTicketError("invalid_ticket", "websocket ticket is invalid, expired, or already used")
        claims = WsTicketClaims.from_row(row)
        if claims.audience != normalized_audience:
            await self._record_event_safe(
                event_type="reject",
                code="audience_mismatch",
                claims=claims,
                client_ip=consume_ip,
                user_agent=consume_user_agent,
            )
            raise WsTicketError("audience_mismatch", "websocket ticket audience mismatch")
        await self._record_event_safe(
            event_type="consume",
            code="ok",
            claims=claims,
            client_ip=consume_ip,
            user_agent=consume_user_agent,
        )
        return claims

    async def record_event(
        self,
        *,
        event_type: str,
        code: str = "",
        audience: str = "",
        subject: str = "",
        role: str = "",
        resource_type: str = "",
        resource_id: str = "",
        site: str = "",
        consume_ip: str = "",
        user_agent: str = "",
    ) -> None:
        claims = WsTicketClaims(
            audience=str(audience or "").strip().lower(),
            subject=str(subject or "").strip().lower(),
            role=str(role or "").strip().lower(),
            resource_type=str(resource_type or "").strip(),
            resource_id=str(resource_id or "").strip(),
            site=str(site or "").strip(),
        )
        await self._record_event_safe(
            event_type=event_type,
            code=code,
            claims=claims,
            client_ip=consume_ip,
            user_agent=user_agent,
        )

    async def _record_event_safe(
        self,
        *,
        event_type: str,
        code: str,
        claims: WsTicketClaims,
        client_ip: str = "",
        user_agent: str = "",
    ) -> None:
        if not self.repository or not hasattr(self.repository, "record_event"):
            return
        if not await self._diagnostics_enabled():
            return
        try:
            await self.repository.record_event(
                event_type=event_type,
                code=code,
                audience=claims.audience,
                subject=claims.subject,
                role=claims.role,
                resource_type=claims.resource_type,
                resource_id=claims.resource_id,
                site=claims.site,
                client_ip=client_ip,
                user_agent=user_agent,
            )
        except Exception as exc:
            if self.logger is not None:
                try:
                    self.logger.debug("[WsTicket] event_record_failed type=%s code=%s err=%s", event_type, code, exc)
                except Exception:
                    pass

    async def _diagnostics_enabled(self) -> bool:
        if self.diagnostics_policy is None:
            return False
        try:
            if hasattr(self.diagnostics_policy, "is_enabled"):
                return bool(await self.diagnostics_policy.is_enabled())
            if callable(self.diagnostics_policy):
                result = self.diagnostics_policy()
                if hasattr(result, "__await__"):
                    result = await result
                return bool(result)
        except Exception as exc:
            if self.logger is not None:
                try:
                    self.logger.debug("[WsTicket] diagnostics_policy_check_failed err=%s", exc)
                except Exception:
                    pass
        return False

    @staticmethod
    def hash_token(token: str) -> str:
        return hashlib.sha256(str(token or "").encode("utf-8")).hexdigest()

    @staticmethod
    def _normalize_required(value: str, field: str) -> str:
        normalized = str(value or "").strip()
        if not normalized:
            raise WsTicketError(f"missing_{field}", f"missing {field}")
        return normalized

File: im_server/internal/ws_ticket/service.go
package ws_ticket

import (
	"context"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
)

type Claims struct {
	Audience     string         `json:"audience"`
	Subject      string         `json:"subject"`
	Role         string         `json:"role"`
	ResourceType string         `json:"resource_type"`
	ResourceID   string         `json:"resource_id"`
	Site         string         `json:"site"`
	Readonly     bool           `json:"readonly"`
	Extra        map[string]any `json:"claims"`
	ExpiresAt    time.Time      `json:"expires_at"`
}

type IssueRequest struct {
	Audience     string
	Subject      string
	Role         string
	ResourceType string
	ResourceID   string
	Site         string
	Readonly     bool
	Extra        map[string]any
	ClientIP     string
	UserAgent    string
}

type IssueResult struct {
	Ticket    string
	TokenHash string
	Claims    Claims
	ExpiresIn int
}

type Service struct {
	db          *pgxpool.Pool
	ttl         time.Duration
	now         func() time.Time
	tokenSize   int
	diagnostics *DiagnosticsPolicyCache
}

func New(db *pgxpool.Pool, ttlSeconds int) *Service {
	if ttlSeconds < 10 {
		ttlSeconds = 45
	}
	if ttlSeconds > 300 {
		ttlSeconds = 300
	}
	return &Service{
		db:          db,
		ttl:         time.Duration(ttlSeconds) * time.Second,
		now:         time.Now,
		tokenSize:   32,
		diagnostics: NewDiagnosticsPolicyCache(db, 5*time.Second),
	}
}

func (s *Service) Issue(ctx context.Context, req IssueRequest) (IssueResult, error) {
	if s == nil || s.db == nil {
		return IssueResult{}, errors.New("ws ticket service unavailable")
	}
	audience := normalizeRequired(req.Audience)
	subject := normalizeRequired(req.Subject)
	if audience == "" || subject == "" {
		return IssueResult{}, errors.New("missing audience or subject")
	}
	token, err := randomToken(s.tokenSize)
	if err != nil {
		return IssueResult{}, err
	}
	tokenHash := HashToken(token)
	now := s.now().Truncate(time.Second)
	expiresAt := now.Add(s.ttl)
	extra := req.Extra
	if extra == nil {
		extra = map[string]any{}
	}
	extraJSON, err := json.Marshal(extra)
	if err != nil {
		extraJSON = []byte("{}")
	}
	claims := Claims{
		Audience:     audience,
		Subject:      subject,
		Role:         strings.ToLower(strings.TrimSpace(req.Role)),
		ResourceType: strings.TrimSpace(req.ResourceType),
		ResourceID:   strings.TrimSpace(req.ResourceID),
		Site:         strings.TrimSpace(req.Site),
		Readonly:     req.Readonly,
		Extra:        extra,
		ExpiresAt:    expiresAt,
	}
	_, err = s.db.Exec(ctx, `
		INSERT INTO ws_tickets (
			token_hash, audience, subject, role, resource_type, resource_id,
			site, readonly, claims, issued_at, expires_at, client_ip, user_agent
		)
		VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9::jsonb, $10, $11, $12, $13)
	`, tokenHash, claims.Audience, claims.Subject, claims.Role, claims.ResourceType, claims.ResourceID,
		claims.Site, claims.Readonly, string(extraJSON), now, expiresAt, trimForStorage(req.ClientIP, 120), trimForStorage(req.UserAgent, 300))
	if err != nil {
		return IssueResult{}, err
	}
	s.recordEvent(ctx, "issue", "ok", claims, req.ClientIP, req.UserAgent)
	return IssueResult{
		Ticket:    token,
		TokenHash: tokenHash,
		Claims:    claims,
		ExpiresIn: int(s.ttl.Seconds()),
	}, nil
}

func (s *Service) Consume(ctx context.Context, ticket string, audience string, clientIP string, userAgent string) (Claims, error) {
	if s == nil || s.db == nil {
		return Claims{}, errors.New("ws ticket service unavailable")
	}
	normalizedTicket := strings.TrimSpace(ticket)
	normalizedAudience := normalizeRequired(audience)
	if normalizedTicket == "" || normalizedAudience == "" {
		s.recordReject(ctx, normalizedAudience, "missing_ticket", clientIP, userAgent)
		return Claims{}, errors.New("missing websocket ticket")
	}
	now := s.now().Truncate(time.Second)
	_, _ = s.db.Exec(ctx, `DELETE FROM ws_tickets WHERE expires_at < $1`, now.Add(-time.Hour))
	var claims Claims
	var extraJSON []byte
	err := s.db.QueryRow(ctx, `
		UPDATE ws_tickets
		SET consumed_at = $2,
		    consume_ip = $3,
		    consume_user_agent = $4
		WHERE token_hash = $1
		  AND audience = $5
		  AND consumed_at IS NULL
		  AND expires_at > $2
		RETURNING audience, subject, role, resource_type, resource_id, site, readonly, claims, expires_at
	`, HashToken(normalizedTicket), now, trimForStorage(clientIP, 120), trimForStorage(userAgent, 300), normalizedAudience).Scan(
		&claims.Audience,
		&claims.Subject,
		&claims.Role,
		&claims.ResourceType,
		&claims.ResourceID,
		&claims.Site,
		&claims.Readonly,
		&extraJSON,
		&claims.ExpiresAt,
	)
	if err != nil {
		s.recordReject(ctx, normalizedAudience, "invalid_ticket", clientIP, userAgent)
		return Claims{}, fmt.Errorf("invalid websocket ticket: %w", err)
	}
	if len(extraJSON) > 0 {
		_ = json.Unmarshal(extraJSON, &claims.Extra)
	}
	if claims.Extra == nil {
		claims.Extra = map[string]any{}
	}
	claims.Audience = strings.TrimSpace(claims.Audience)
	claims.Subject = strings.ToLower(strings.TrimSpace(claims.Subject))
	claims.Role = strings.ToLower(strings.TrimSpace(claims.Role))
	s.recordEvent(ctx, "consume", "ok", claims, clientIP, userAgent)
	return claims, nil
}

func (s *Service) recordReject(ctx context.Context, audience string, code string, clientIP string, userAgent string) {
	s.recordEvent(ctx, "reject", code, Claims{Audience: strings.TrimSpace(audience)}, clientIP, userAgent)
}

func (s *Service) recordEvent(ctx context.Context, eventType string, code string, claims Claims, clientIP string, userAgent string) {
	if s == nil || s.db == nil {
		return
	}
	if s.diagnostics == nil || !s.diagnostics.Enabled(ctx) {
		return
	}
	_, _ = s.db.Exec(ctx, `
		INSERT INTO ws_ticket_events (
			event_type, code, audience, subject, role, resource_type,
			resource_id, site, client_ip, user_agent, created_at
		)
		VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11)
	`,
		trimForStorage(eventType, 40),
		trimForStorage(code, 80),
		trimForStorage(claims.Audience, 40),
		trimForStorage(claims.Subject, 120),
		trimForStorage(claims.Role, 40),
		trimForStorage(claims.ResourceType, 80),
		trimForStorage(claims.ResourceID, 160),
		trimForStorage(claims.Site, 80),
		trimForStorage(clientIP, 120),
		trimForStorage(userAgent, 300),
		s.now().Truncate(time.Second),
	)
}

func HashToken(token string) string {
	sum := sha256.Sum256([]byte(strings.TrimSpace(token)))
	return hex.EncodeToString(sum[:])
}

func randomToken(size int) (string, error) {
	buf := make([]byte, size)
	if _, err := rand.Read(buf); err != nil {
		return "", err
	}
	return base64.RawURLEncoding.EncodeToString(buf), nil
}

func normalizeRequired(value string) string {
	return strings.ToLower(strings.TrimSpace(value))
}

func trimForStorage(value string, limit int) string {
	normalized := strings.TrimSpace(value)
	if limit > 0 && len(normalized) > limit {
		return normalized[:limit]
	}
	return normalized
}
